File: day25/day25.py
SNAFU = {'1': 1, '2': 2, '0': 0, '-': -1, '=': -2}
SNAFU_INV = {0: ('0', 0), 1: ('1', 0), 2: ('2', 0), 3: ('=', 1), 4: ('-', 1), 5: ('0', 1)}

def to_dec(s):
    return sum([SNAFU[c] * (5**i) for i, c in enumerate(s[::-1])])

def to_snafu(n):
    snafu = ''
    extra = 0
    while n > 0:
        symbol, extra = SNAFU_INV[n % 5 + extra]
        snafu = symbol + snafu
        n //= 5
    if extra:
        snafu = SNAFU_INV[extra][0] + snafu
    return snafu

def main():
    with open('input.txt') as f:
        data = f.readlines()

    lines = [line.strip() for line in data]
    print(to_snafu(sum([to_dec(line) for line in lines])))

if __name__ == '__main__':
    main()

File: day25/day25.rs
use std::collections::HashMap;

fn to_dec(s: &str, snafu: &HashMap<char, isize>) -> isize {
    s.chars()
        .rev()
        .enumerate()
        .map(|(p, c)| snafu.get(&c).unwrap() * 5_isize.pow(p as u32))
        .sum()
}

fn to_snafu(n: isize, snafu_inv: [(&str, isize); 6]) -> String {
    let mut snafu = "".to_string();
    let mut m = n;
    let mut extra = 0;
    while m > 0 {
        extra = snafu_inv[(m % 5 + extra) as usize].1;
        snafu.push_str(snafu_inv[(m % 5 + extra) as usize].0);
        m = m / 5;
    }
    if extra > 0 {
        snafu.push_str(snafu_inv[extra as usize].0);
    }
    return snafu.chars().rev().collect();
}

fn main() {
    let raw_data = std::fs::read_to_string("input.txt").unwrap();
    let snafu: HashMap<char, isize> = HashMap::from([('1', 1), ('2', 2), ('0', 0), ('-', -1), ('=', -2)]);
    let snafu_inv: [(&str, isize); 6] = [("0", 0), ("1", 0), ("2", 0), ("=", 1), ("-", 1), ("0", 1)];
    println!( "{}", to_snafu(raw_data.split("\n").map(|x| to_dec(x, &snafu)).sum(),snafu_inv));
}
